auth_mellon_util.c

```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
+ *                              may be NULL. Bitmask of option flags.
  *
  * Returns:
- *   true if the PAOS header is valid, false otherwise
+ *   true if the PAOS header is valid, false otherwise. If options is non-NULL
+ *   then the set of option flags is returned there.
  *
  */
-#define PAOS_VERSION_TOKEN "ver=\"" LASSO_PAOS_HREF "\""
-#define ECP_SERVICE_TOKEN  "\"" LASSO_ECP_HREF "\""
-bool am_validate_paos_header(request_rec *r, const char *header)
+bool am_parse_paos_header(request_rec *r, const char *header,
+                             ECPServiceOptions *options_return)
 {
     bool result = false;
-    char **semicolon_tokens = NULL;
-    char *token = NULL;
-    guint len;
+    ECPServiceOptions options = 0;
+    apr_array_header_t *tokens;
+    apr_size_t i;
+    char *error;
 
-    if (header == NULL) {
-        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
-                     "invalid PAOS header, NULL");
+    ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
+                  "PAOS header: \"%s\"", header);
+
+    tokens = tokenize(r->pool, header, true, &error);
+
+#ifdef DEBUG
+    dump_tokens(r, tokens);
+#endif
+
+    if (error) {
+        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, error);
         goto cleanup;
     }
 
-    /* Split the header on the semicolon character */
-    semicolon_tokens = g_strsplit(header, ";", 0);
-
-    /* There must be exactly two tokens after splitting */
-    len = g_strv_length(semicolon_tokens);
-    if (len != 2) {
+    /* Header must begin with "ver=xxx" where xxx is paos version */
+    if (!is_token(tokens, 0, TOKEN_IDENTIFIER, "ver") ||
+        !is_token(tokens, 1, TOKEN_EQUAL, NULL) ||
+        !is_token(tokens, 2, TOKEN_DBL_QUOTE_STRING, LASSO_PAOS_HREF)) {
         ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
-                     "invalid PAOS header, "
-                     "expected 2 tokens seperated by semicolon, header=\"%s\"",
-                     header);
+                      "invalid PAOS header, "
+                      "expected header to begin with ver=\"%s\", "
+                      "actual header=\"%s\"",
+                      LASSO_PAOS_HREF, header);
         goto cleanup;
+
     }
 
-    /* Validate the first token as the PAOS version */
-    token = g_strstrip(semicolon_tokens[0]);
-    if (!g_str_equal(token, PAOS_VERSION_TOKEN)) {
+    /* Next is the service value, separated from the version by a semicolon */
+    if (!is_token(tokens, 3, TOKEN_SEMICOLON, NULL)) {
         ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                      "invalid PAOS header, "
-                     "expected first token to be \"%s\", "
-                     "but found \"%s\" in header=\"%s\"",
-                     PAOS_VERSION_TOKEN, token, header);
+                     "expected semicolon after PAOS version "
+                     "but found \"%s\" at position %lu in header=\"%s\"",
+                      APR_ARRAY_IDX(tokens, 3, Token).str,
```

Comment:
Here you read beyond the end of the array if the tokenized string ends after the double-quoted version. I.e.:
```
PAOS: ver="urn:liberty:paos:2003-08"
```